qiskit/pulse/instructions/instruction.py

```diff
@@ -221,34 +221,36 @@ def append(self, schedule: ScheduleComponent,
 
     def assign_parameters(self,
                           value_dict: Dict[ParameterExpression,
-                                           Union[ParameterExpression, int, float, complex]],
-                          inplace: bool = True) -> 'Instruction':
-        """Assign the parameters in this instruction according to the input.
+                                           Union[ParameterExpression, int, float, complex]]
+                          ) -> 'Instruction':
+        """Modify and return self with parameters assigned according to the input.
 
         Args:
             value_dict: A mapping from Parameters to either numeric values or another
                 Parameter expression.
-            inplace: Modify this ``Instruction`` iff True, else return a new ``Instruction``.
 
         Returns:
-            Instruction with updated parameters (a new one if not inplace, otherwise self).
+            Self with updated parameters.
         """
         new_operands = list(self.operands)
+
         for idx, op in enumerate(self.operands):
             for parameter, value in value_dict.items():
                 if isinstance(op, ParameterExpression) and parameter in op.parameters:
                     new_operands[idx] = new_operands[idx].assign(parameter, value)
                 elif (isinstance(op, Channel)
                       and isinstance(op.index, ParameterExpression)
                       and parameter in op.index.parameters):
-                    # TODO: Check int type?
                     new_index = new_operands[idx].index.assign(parameter, value)
+                    if not new_index.parameters:
```

Comment:
The implications of this are interesting. How does this work if there is an are two overlapping instructions with the same parameter, would these not be invalid? Will this be resolved at the binding time?

I am somewhat uncomfortable perhaps because I am unclear with the Qiskit parameter implementation of what is supposed to be macros for code-generation time and what is imagined to be represented as reusable functions/code to be targeted to the backend hardware and executed by the QPU. 

As an analogy to a classical compiler, this is akin to a virtual register, pre-register allocation, but post-scheduling. During compilation, it is fine to re-write registers. However, post-compilation within a compiled program register locations will not be passed as values with resolved scheduling (which is very critical), they will be fixed in the binary and not evaluable. We cannot hope to dynamically at runtime rewrite channels within functions, except by ahead of time monomorphization of these functions.

Reply:
Resolved at assignment time. I'll add a test for it